urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/server/AbstractINDIServer.java

```diff
@@ -0,0 +1,812 @@
+   */
+  public synchronized void destroyNetworkDriver(String host, int port) {
+    String networkName = host + ":" + port;
+
+    print("Removing network driver " + networkName);
+
+    destroyIdentifiedDrivers(networkName);
+  }
+
+  /**
+   * Removes a Device from the list of devices. Called by the Devices to be
+   * removed when connection brokes
+   *
+   * @param device The Device to be removed.
+   */
+  protected void removeDevice(INDIDevice device) {
+    String[] names = device.getNames();
+
+    devices.remove(device);
+    clients.remove(device);
+
+    notifyClientsDeviceRemoved(names);
+
+    driverDisconnected(device.getDeviceIdentifier(), names);
+  }
+
+  /**
+   * Notifies the listening clients that some particular Devices have been
+   * removed by sending
+   * <code>delProperty</code> messages.
+   *
+   * @param deviceNames The names of the Devices that have been removed.
+   */
+  private void notifyClientsDeviceRemoved(String[] deviceNames) {
+    for (int h = 0 ; h < deviceNames.length ; h++) {
+      String deviceName = deviceNames[h];
+
+      String message = "<delProperty device=\"" + deviceName + "\" />";
+
+      ArrayList<INDIDeviceListener> list = this.getClientsListeningToDevice(deviceName);
+
+      for (int i = 0 ; i < list.size() ; i++) {
+        INDIDeviceListener c = list.get(i);
+
+        c.sendXMLMessage(message);
+      }
+
+      ArrayList<INDIDeviceListener> list2 = this.getClientsListeningToSingleProperties(deviceName);
+
+      for (int i = 0 ; i < list2.size() ; i++) {
+        INDIDeviceListener c = list2.get(i);
+
+        c.sendXMLMessage(message);
+      }
+    }
+  }
+
+  /**
+   * Removes a Client from the List of clients. Called by the clients when the
+   * connection is broken.
+   *
+   * @param client The Client to remove.
+   */
+  protected void removeClient(INDIClient client) {
+    clients.remove(client);
+
+    connectionWithClientBroken(client);
+  }
+
+  /**
+   * Checks if a Class is subclass of INDIDriver.
+   *
+   * @param c The class to check
+   * @return
+   * <code>true</code> if the Class inherits from INDIDriver.
+   * <code>false</code> otherwise.
+   */
+  private boolean isINDIDriver(Class c) {
+if (INDINotLoadableDriver.class.isAssignableFrom(c)) {
+  return false; 
+}
+    Class s = c.getSuperclass();
+
+    while (s != null) {
+      if (s == INDIDriver.class) {
+        return true;
+      }
+
+      s = s.getSuperclass();
+    }
+
+    return false;
+  }
+
+  /**
+   * Gets a list of Strings with the name of the classes in a JAR file.
+   *
+   * @param file The JAR file.
+   * @return A list with all the names of the classes in the JAR file.
+   * @throws INDIException if there is any problem accessing to the JAR file.
+   */
+  private ArrayList<String> getClassesInJAR(String file) throws INDIException {
+
+    ArrayList<String> list = new ArrayList<String>();
+
+    try {
+      JarInputStream jarFile = new JarInputStream(new FileInputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *jarFile*. Execution paths that do not contain closure statements were detected. Ensure *jarFile* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).